src/Symfony/Component/Routing/Loader/ObjectLoader.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Routing\Loader;
+
+use Symfony\Component\Config\Loader\Loader;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\Routing\RouteCollection;
+
+/**
+ * A route loader that calls a method on an object to load the routes.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+abstract class ObjectLoader extends Loader
+{
+    /**
+     * @deprecated since Symfony 4.4 and will be removed in 5.0, implement and use "getObject" instead.
+     */
+    abstract protected function getServiceObject($id);
+
+    /**
+     * Returns the object that the method will be called on to load routes.
+     *
+     * For example, if your application uses a service container,
+     * the $id may be a service id.
+     *
+     * @return object
+     */
+    /*abstract in Symfony 5.0 */
+    protected function getObject(string $id)
+    {
+        @trigger_error(sprintf('The "%s::getServiceObject()" method is deprecated since Symfony 4.4 and will be removed in 5.0, implement and use "%s" instead.', self::class, __FUNCTION__), E_USER_DEPRECATED);
+
+        return $this->getServiceObject($id);
+    }
+
+    /**
+     * Calls the object method that will load the routes.
+     *
+     * @param string      $resource object_id::method
+     * @param string|null $type     The resource type
+     *
+     * @return RouteCollection
+     */
+    public function load($resource, $type = null)
+    {
+        if (!preg_match('/^[^\:]+(?:::?(?:[^\:]+))?$/', $resource)) {
+            throw new \InvalidArgumentException(sprintf('Invalid resource "%s" passed to the %s route loader: use the format "object_id::method" or "object_id" if your object class has an "__invoke" method.', $resource, \is_string($type) ? '"'.$type.'"' : 'object'));
+        }
+
+        if (1 === substr_count($resource, ':')) {
+            $resource = str_replace(':', '::', $resource);
+            @trigger_error(sprintf('Referencing object route loaders with a single colon is deprecated since Symfony 4.1. Use %s instead.', $resource), E_USER_DEPRECATED);
```

Comment:
moving this check to the deprecated child class could make sense, having all deprecated code in one place makes it easier to remove in the next major.

Reply:
That would make sense but it's hard since the check is after the argument check, and I guess we don't want to trigger the deprecation if the argument is invalid. So : 
```php
if ($useSingleColon = (1 === substr_count($resource, ':'))) {
            $resource = str_replace(':', '::', $resource);
        }

        try {
            return parent::load($resource, $type);
        } catch (\InvalidArgumentException $e) {
        } finally {
            if ($useSingleColon && !isset($e)) {
                @trigger_error(sprintf('Referencing object route loaders with a single colon is deprecated since Symfony 4.1. Use %s instead.', $resource), E_USER_DEPRECATED);
            }
        }

Also, when we remove this deprecation in 5.0, I guess we actually have to modify the above regex to throw an invalid argument exception if ":" is used.

So I'm not sure if it will be easier 🤔

Reply:
can't we duplicate the argument check in the child class?

Reply:
Yes let's do it.

Also since `ObjectLoader` is a new class, I can impose the `::` usage right? So the check is actually not the same.